### ISSUE TITLE
trove4j is LGPL-licensed

### ISSUE THREAD

**Issue by HansBrende** [owlcs/owlapi#774]
The [trove4j](https://mvnrepository.com/artifact/net.sf.trove4j/trove4j/3.0.3) dependency is LGPL-licensed, which is not compatible with the Apache license.

See [ANY23-372](https://issues.apache.org/jira/browse/ANY23-372).

On the **version5** branch, the trove4j library is used in the following 7 classes:
```java
// api/src/main/java/org/semanticweb/owlapi/io/RDFTriple
import gnu.trove.map.hash.TObjectIntHashMap;

// impl/src/main/java/uk/ac/manchester/cs/owl/owlapi/MapPointer
import gnu.trove.map.hash.THashMap;
import gnu.trove.set.hash.THashSet;

// parsers/src/main/java/org/semanticweb/owlapi/rdf/model/RDFTranslator
import gnu.trove.map.custom_hash.TObjectIntCustomHashMap;
import gnu.trove.strategy.IdentityHashingStrategy;

// ONLY USED IN TESTS:
// contract/src/test/java/org/semanticweb/owlapi/api/test/baseclasses/PlainLiteralTypeFoldingAxiomSet
import gnu.trove.set.hash.TCustomHashSet;
import gnu.trove.strategy.HashingStrategy;

// ONLY USED IN TESTS:
// tools/src/main/java/uk/ac/manchester/cs/atomicdecomposition/AtomicDecompositionImpl
import gnu.trove.map.hash.TObjectIntHashMap;

// UNUSED CLASS:
// tools/src/main/java/uk/ac/manchester/cs/chainsaw/ArrayIntMap (class never used in project)
import gnu.trove.list.TIntList;
import gnu.trove.list.array.TIntArrayList; 

// UNUSED PACKAGE IMPORT:
// oboformat/src/main/java/org/obolibrary/oboformat/writer/OBOFormatWriter
import gnu.trove.map.hash.TObjectIntHashMap; //unused import
```
Note that the trove4j dependency only occurs in `OBOFormatWriter` as an unused package import, `ArrayIntMap` is never used anywhere in the project, and `AtomicDecompositionImpl` and `PlainLiteralTypeFoldingAxiomSet` are only used in the unit tests.

On the **master** branch, the trove4j library is only used in 2 classes:
```java
// impl/src/main/java/uk/ac/manchester/cs/owl/owlapi/MapPointer
import gnu.trove.map.hash.THashMap;
import gnu.trove.set.hash.THashSet;

// ONLY USED IN TESTS:
// contract/src/test/java/org/semanticweb/owlapi/api/test/baseclasses/PlainLiteralTypeFoldingAxiomSet
import gnu.trove.set.hash.TCustomHashSet;
import gnu.trove.strategy.HashingStrategy;
```

**Comment by HansBrende** [comment]
I'd recommend [java.util.HashSet](https://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html), [java.util.HashMap](https://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html), and [java.util.IdentityHashMap](https://docs.oracle.com/javase/7/docs/api/java/util/IdentityHashMap.html) as replacements for trove4j. 

Or, if you're really speed-conscious, [fastutil](https://github.com/vigna/fastutil) is an option.

**Comment by ignazio1977** [comment]
Unfortunately `MapPointer` is central in holding `OWLOntology` data. The standard Java collections aren't suitable for the job - we know that because of the speed and memory improvements we saw when we switched from them to Trove.

Fastutil looks good, however its jar is 17 megabytes. This is twice the size of the owlapi-osgidistribution jar, so it feels excessive.

According to this [benchmark](http://java-performance.info/large-hashmap-overview-jdk-fastutil-goldman-sachs-hppc-koloboke-trove/) [HPCC](https://github.com/carrotsearch/hppc) is slightly faster, and its jar comes in at 1.2 megabytes, half the size of trove4j. I haven't looked in detail to the classes available, but it feels like the best choice at the moment.

**Comment by HansBrende** [comment]
@ignazio1977 Also if you did want to go the fastutil route, their [README](https://github.com/vigna/fastutil/blob/master/README.md) contains the following section:

> Note that the jar file is huge, due to the large number of classes. To create a small, customized fastutil jar (which you can put in your repo, local maven, etc.), we provide the find-deps.sh shell script. It has mild prerequisites, as only the jdeps tool is required (bundled with JDK 8). It can be used to identify all fastutil classes your project uses and build a minimized jar only containing the necessary classes. You can also have a look at AutoJar or similar tools to extract automatically the necessary classes.

**Comment by HansBrende** [comment]
@ignazio1977 oh and FYI: the benchmark you linked to appears to have been outdated by [this one](http://java-performance.info/hashmap-overview-jdk-fastutil-goldman-sachs-hppc-koloboke-trove-january-2015/).

**Comment by ignazio1977** [comment]
Yes that's the benchmark I meant to link. I've read both but pasted the wrong URL here.

**Comment by HansBrende** [comment]
@ignazio1977 So I'm a little confused then about what you said about HPCC being the fastest, because the `MapPointer` class uses an Object-to-Object mapping, and according to the revised benchmark, `java.util.HashMap` outperforms all other Object-to-Object Map implementations for `get` operations, and outperforms HPCC for both `put` operations and `remove` operations. So if you were going to go the HPCC route, you might as well just use the jdk `HashMap` instead.

**Comment by HansBrende** [comment]
@ignazio1977 
... from that benchmark, it looks to me like the best performers (the ones better than the jdk) were: (1) FastUtil, (2) Koloboke (a.k.a. Hftc), and (3) Eclipse Collections (a.k.a. GS Collections). Of those, only FastUtil and Eclipse Collections seem to be actively maintained. If you go the FastUtil route, I'd recommend renaming the packages you need to avoid conflicting class definitions down the line (e.g., `org.semanticweb.shaded.it.unimi.dsi.fastutil.objects.Object2ObjectMap`).

**Comment by ignazio1977** [comment]
Regarding the standard collections, the main problem is not speed but memory footprint. I've got an HPPC-RT (a fork of HPPC with focus on fewer object creations on iterations) test locally and compared with Trove and standard collections. Even on object to object maps, both alternatives are better than the standard libraries (The ontologies I've used for the test are full-galen, gazetteer and NCBI taxon).

In terms of speed, the libraries are all comparable - within a few percents. HPPC-RT gains 5% on Trove and little more on the standard collections. However, in memory usage, the standard collections use more memory from a few megabytes for Gazetteer to one extra gigabyte for NCBI taxon.

HPPC-RT last release was in 2017, so, while not as actively developed as fastutil, it is still undergoing some development.

Koloboke compile offers a very interesting alternative, allowing to build classes that are tailored to our needs. That's something I intend to look into; however it introduces large runtime dependencies out of the box. It would require the same repackaging work to reduce the size of the release jars, and I'm trying to understand if including only part of another library has effects on the licensing.

Short term, HPPC-RT looks like the best choice - the performance change is minimal, the licensing is straightforward, the size of the release jars decreases slightly. There are issues raised that might cause MapPointer to be rewritten; at that point we might decide to change strategy.

**Comment by HansBrende** [comment]
@ignazio1977 ah, that makes sense. In that case, any idea when the next owlapi maven artifact will be released? We're pushing out a new Any23 release soon and it would be nice to have license-compatible dependencies. Ditto, I am sure, for the Nutch devs.

**Comment by ignazio1977** [comment]
Soon as I can cherry pick the change to version 4 and 5.

**Comment by ignazio1977** [comment]
4.5.4 released, 5.1.6 released
